File: app/src/main/java/com/supercilex/robotscouter/ui/scout/ScoutCounterValueDialog.java
package com.supercilex.robotscouter.ui.scout;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.text.InputType;

import com.google.firebase.database.DatabaseReference;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.util.DatabaseHelper;

import java.math.BigDecimal;

public class ScoutCounterValueDialog extends ScoutValueDialogBase<Integer> {
    private static final String TAG = "ScoutCounterValueDialog";

    public static void show(FragmentManager manager, DatabaseReference ref, String currentValue) {
        ScoutValueDialogBase dialog = new ScoutCounterValueDialog();

        Bundle args = DatabaseHelper.getRefBundle(ref);
        args.putString(CURRENT_VALUE, currentValue);
        dialog.setArguments(args);

        dialog.show(manager, TAG);
    }

    @Override
    public void onShow(DialogInterface dialog) {
        super.onShow(dialog);
        mValue.setInputType(InputType.TYPE_CLASS_NUMBER);
    }

    @Override
    protected boolean onClick() {
        try {
            new BigDecimal(mValue.getText().toString()).intValueExact();
            return super.onClick();
        } catch (NumberFormatException | ArithmeticException e) {
            mInputLayout.setError(getString(R.string.invalid_team_number));
            return false;
        }
    }

    @Override
    protected Integer getValue() {
        return Integer.valueOf(mValue.getText().toString());
    }

    @Override
    protected int getTitle() {
        return R.string.edit_value;
    }

    @Override
    protected int getHint() {
        return R.string.value;
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scout/viewholder/ScoutCounterValueDialog.kt
package com.supercilex.robotscouter.ui.scout.viewholder

import android.content.DialogInterface
import android.support.v4.app.FragmentManager
import android.text.InputType
import com.google.firebase.database.DatabaseReference
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.ui.scout.ScoutValueDialogBase
import com.supercilex.robotscouter.util.DatabaseHelper
import com.supercilex.robotscouter.util.isNumber
import com.supercilex.robotscouter.util.show

class ScoutCounterValueDialog : ScoutValueDialogBase<Long>() {
    override val value get() = lastEditText.text.toString().toLong()
    override val title = R.string.edit_value
    override val hint = R.string.value

    override fun onShow(dialog: DialogInterface) {
        super.onShow(dialog)
        lastEditText.inputType = InputType.TYPE_CLASS_NUMBER
    }

    override fun onAttemptDismiss() = if (lastEditText.text.toString().isNumber()) {
        super.onAttemptDismiss()
    } else {
        inputLayout.error = getString(R.string.number_too_big_error); false
    }

    companion object {
        private const val TAG = "ScoutCounterValueDialog"

        fun show(manager: FragmentManager, ref: DatabaseReference, currentValue: String) =
                ScoutCounterValueDialog().show(manager, TAG, DatabaseHelper.getRefBundle(ref)) {
                    putString(CURRENT_VALUE, currentValue)
                }
    }
}
